[17.0][ADD] l10n_es_aeat_sii_invoice_summary_pos

edescalona opened the pull request:
@BinhexTeam

It allows you to fill in the first and second invoice fields using a button, according to the date defined by the user, in the SII invoice tab.

edescalona commented:
Ready for review

christian-ramos-tecnativa commented on l10n_es_aeat_sii_invoice_summary_pos/models/account_move.py:
```diff
@@ -0,0 +1,105 @@
+            amount_total = sum(
+                [pos_order["amount_total"] for pos_order in pos_order_ids]
+            )
+            sii_line = self.invoice_line_ids.filtered(lambda x: x.is_sii_line)
```
En lugar de utilizar este nuevo campo a nivel de apunte, considero mejor simplemente borrar todas las líneas y regenerarlas cada vez que se llama a este método

edescalona replied:
Hola @Christian-RB pudiera ser, pero mi idea era por si el usuario agregaba lineas manualmente no hacer que las vuelva a agregar, solo actualizar la linea referente al SII.

christian-ramos-tecnativa replied:
En ese caso considero que cualquier modificación debería hacerse a posteriori de la generación, por lo tanto lo dejaría bien redactado a nivel de README y con eso debería ser suficiente. En principio este tipo de factura, se presta a tener una línea solo con el monto total

edescalona replied:
ok @Christian-RB gracias

christian-ramos-tecnativa commented on l10n_es_aeat_sii_invoice_summary_pos/models/account_move.py:
```diff
@@ -0,0 +1,105 @@
+            invoice._valid_sii_dates()
+            invoice._populate_invoice_summary_by_dates()
+
+    def action_post(self):
+        res = super().action_post()
+        for invoice in self:
+            if invoice.is_invoice_summary:
+                invoice.populate_invoice_summary_by_dates()
+        return res
```
No creo que sea buena idea realizar este recálculo justo en el momento de validar la factura. Principalmente porque se puede dar el caso en el que el usuario ha generado la factura y por el motivo que sea ahora existen pos.orders distintas o quiere modificar a posteriori algún valor. Con esto el usuario nunca estará 100% seguro de lo que está validando.
```suggestion
```

edescalona replied:
Hola @Christian-RB vale

christian-ramos-tecnativa commented on l10n_es_aeat_sii_invoice_summary_pos/models/account_move.py:
```diff
@@ -0,0 +1,105 @@
+
+    sii_start_date = fields.Date("Start Date")
+    sii_end_date = fields.Date("End Date")
+    sii_tickets = fields.Text()
```
En lugar de un campo de texto creo que quedaría mejor un many2one en las pos.order `invoice_summary` que apunte a la factura que se presentó, de esta forma podríamos añadir un botón que despliegue las órdenes seleccionadas. Habría que añadir al recomputar desenlazar los pedidos que no correspondan.

edescalona replied:
ok @Christian-RB

christian-ramos-tecnativa reviewed (changes requested):
Revisa mis comentarios y dime que te parece. Gracias

christian-ramos-tecnativa commented on l10n_es_aeat_sii_invoice_summary_pos/models/account_move.py:
```diff
@@ -0,0 +1,95 @@
+        )
+        for set_pos_order in set_pos_order_ids:
+            if set_pos_order not in pos_order_ids:
+                set_pos_order_ids.invoice_summary_id = False
```
Esto debería hacer exactamente lo mismo, es código sin probar, ten cuidado
```suggestion
        (set_pos_order_ids - pos_order_ids).write({"invoice_summary_id": False})
```

edescalona committed:
[ADD] l10n_es_aeat_sii_invoice_summary_pos: Rellenar automáticamente los campos del Resumen de Factura Simplificada SII.

edescalona commented:
Hola @Christian-RB listo para revisión, gracias.

christian-ramos-tecnativa commented on l10n_es_aeat_sii_invoice_summary_pos/models/account_move.py:
```diff
@@ -0,0 +1,95 @@
+
+        if pos_order_ids:
+            self.set_order_summary(pos_order_ids)
+            for pos_order in pos_order_ids:
+                pos_order.invoice_summary_id = self.id
```
```suggestion
            pos_order_ids.write({"invoice_summary_id": self.id})
```

christian-ramos-tecnativa commented on l10n_es_aeat_sii_invoice_summary_pos/models/account_move.py:
```diff
@@ -0,0 +1,95 @@
+
+            self.sii_invoice_summary_start = pos_order_ids[0].name
+            self.sii_invoice_summary_end = pos_order_ids[-1].name
+            amount_total = sum([pos_order.amount_total for pos_order in pos_order_ids])
+            self.invoice_line_ids = [
+                Command.clear(),
+                Command.create(
+                    {
+                        "name": "{}-{}".format(
+                            self.sii_invoice_summary_start,
+                            self.sii_invoice_summary_end,
+                        ),
+                        "price_unit": amount_total,
+                    }
+                ),
+            ]
```
Esto hay que afinarlo, no podemos coger el monto total, ya que debemos tener una línea por combinación de impuestos. Ejemplo:
Si tengo 5 pedidos, todos ellos con una línea con la siguiente distribución:
Taxes       | Total
21%,10% | 100
21%          | 200
10%          | 300
10%          | 400
21%,3%    | 500
Esto debería resultar en una factura con las siguientes líneas:
21%,10% | 100
21%          | 200
10%          | 700
21%,3%    | 500
Déjame saber si me he explicado correctamente

edescalona committed:
[REF] l10n_es_aeat_sii_invoice_summary_pos: Generar lineas de facturas agrupadas por los impuestos.

[IMP] l10n_es_aeat_sii_invoice_summary_pos: Valor por defecto para el campo Facturas simplificadas en la AEAT en los partner

[IMP] l10n_es_aeat_sii_invoice_summary_pos: Validar campos de primer y última factura cuando no sea draft

[IMP] l10n_es_aeat_sii_invoice_summary_pos: Valor por defecto para el campo Facturas simplificadas en la AEAT en el partner Simplified Invoice Partner

christian-ramos-tecnativa commented:
Vas a tener que volver a empujar la rama igualmente a ver si el CI sigue fallando

edescalona commented:
listo @christian-ramos-tecnativa

pedrobaeza reviewed (commented):
No termino de entender la utilidad de esto, ya que así se estaría generando según entiendo doble contabilidad: la de la sesión al cerrar, y luego la de la factura de resumen, ¿o hay algo que se me escapa?

christian-ramos-tecnativa commented:
Ese problema ya lo presenta el módulo principal https://github.com/OCA/l10n-spain/tree/17.0/l10n_es_aeat_sii_invoice_summary, este lo único que pretende es tener una forma rápida de crear esas facturas agrupadas utilizando las ventas del tpv

pedrobaeza commented:
No, porque llevan finalidades distintas. El otro módulo pretende cubrir el hueco cuando el TPV no está en Odoo (el típico paquete de hardware + software de TPV que te venden y que utilizas tal cual, o N razones para no utilizar el de Odoo), y entonces solo introduces la contabilidad vía esta factura resumen, pero insisto, si el TPV es en Odoo, entonces con esto sí que tendrías doble contabilidad.

christian-ramos-tecnativa commented:
Estás acertado con el tema de la contabilidad doble y que usando este módulo tendrás que cancelar esos asientos o compensarlos de alguna manera, pero hay empresas que presentan estas facturas agrupadas por mes y lo normal es que las sesiones de odoo sean diarias. Por lo que este módulo viene a resolver ese punto sin necesidad de crearlo manualmente.

pedrobaeza commented:
La solución para ese caso debe ser enganchar el `pos.session` al `sii.mixin`, y enviar ese resumen desde ahí, no este módulo. Lo siento, pero no veo conveniente fusionar un módulo que promueve unas prácticas incorrectas.
